src/java/main/org/apache/zookeeper/client/StaticHostProvider.java

```diff
@@ -113,5 +124,21 @@ public InetSocketAddress next(long spinDelay) {
 
     public void onConnected() {
         lastIndex = currentIndex;
+        connectedSinceNext = true;
```

Comment:
Should this be an atomic operation?

Reply:
Both `next()` and `onConnected()` are called from the same thread sequentially otherwise yeah it'd need to be atomic or volatile.

Reply:
Is this really guaranteed? `next()` and `onConnected()` are both public methods, and as such we may be making assumptions on the clients of this class that are not adhered to. Not a blocker from my perspective though.

Reply:
It is not strictly guaranteed, but within the context of the ZK codebase they are guaranteed to be called sequentially in a single thread.

Reply:
(Same Q was asked on the Apache PR as well)